src/GafferVDB/PointsGridToPoints.cpp

```diff
@@ -0,0 +1,557 @@
+	},
+	{
+		openvdb::typeNameAsString<openvdb::Mat4d>(),
+		Functions(
+			[](size_t size) -> IECore::DataPtr { return createArray<Imath::M44d>(size); },
+			[](IECore::Data *destArray, const openvdb::points::AttributeArray& array, LeafIter leafIter)  { appendData<Imath::M44d, openvdb::Mat4d>( destArray, array, leafIter ); }
+		)
+	},
+
+	// quaternions - single & double
+	{
+		openvdb::typeNameAsString<openvdb::math::Quats>(),
+		Functions(
+			[](size_t size) -> IECore::DataPtr { return createArray<Imath::Quatf>(size); },
+			[](IECore::Data *destArray, const openvdb::points::AttributeArray& array, LeafIter leafIter)  { appendData<Imath::Quatf, openvdb::math::Quats>( destArray, array, leafIter ); }
+		)
+	},
+	{
+		openvdb::typeNameAsString<openvdb::math::Quatd>(),
+		Functions(
+			[](size_t size) -> IECore::DataPtr { return createArray<Imath::Quatd>(size); },
+			[](IECore::Data *destArray, const openvdb::points::AttributeArray& array, LeafIter leafIter)  { appendData<Imath::Quatd, openvdb::math::Quatd>( destArray, array, leafIter ); }
+		)
+	},
+};
+
+void appendPrimitiveVariableData( IECoreScene::PrimitiveVariableMap &variableMap,
+	const std::string &name,
+	const std::string &type,
+	LeafIter leafIter,
+	const openvdb::points::AttributeArray &arrayData,
+	uint64_t count)
+{
+
+	auto itConverter = converters.find( type );
+	if ( itConverter == converters.end() )
+	{
+		return;
+	}
+
+	IECoreScene::PrimitiveVariable primVar;
+
+	auto primVarIt = variableMap.find( name );
+	if( primVarIt == variableMap.end() )
+	{
+		primVar = IECoreScene::PrimitiveVariable( IECoreScene::PrimitiveVariable::Vertex, itConverter->second.m_create( count ) );
+		variableMap[name] = primVar;
+	}
+	else
+	{
+		primVar = primVarIt->second;
+	}
+
+	itConverter->second.m_append( primVar.data.get(), arrayData, leafIter);
+}
+
+IECoreScene::PointsPrimitivePtr createPointsPrimitive( openvdb::GridBase::ConstPtr baseGrid, std::function<bool( const std::string & )> primitiveVariableFilter )
+{
+	openvdb::points::PointDataGrid::ConstPtr pointsGrid = openvdb::GridBase::constGrid<openvdb::points::PointDataGrid>( baseGrid );
+	if( !pointsGrid )
+	{
+		return nullptr;
+	}
+
+	openvdb::Index64 count = openvdb::points::pointCount( pointsGrid->tree() );
+
+	IECore::V3fVectorDataPtr pointData = new IECore::V3fVectorData();
+	auto &points = pointData->writable();
+	points.reserve( count );
+
+	IECoreScene::PrimitiveVariableMap primVars;
+
+	for( auto leafIter = pointsGrid->tree().cbeginLeaf(); leafIter; ++leafIter )
+	{
+		const openvdb::points::AttributeArray &array = leafIter->constAttributeArray( "P" );
+		openvdb::points::AttributeHandle<openvdb::Vec3f> positionHandle( array );
+
+		for( auto indexIter = leafIter->beginIndexOn(); indexIter; ++indexIter )
+		{
+			const openvdb::points::AttributeSet &attributeSet = leafIter->attributeSet();
+			const openvdb::points::AttributeSet::Descriptor &descriptor = attributeSet.descriptor();
+
+			for (const auto &it : descriptor.map() )
```

Comment:
So I have to get the first leaf's descriptor? Or is there another API which implies this constant over all leafs?

Reply:
That's what we recommend, just get the first leaf descriptor and assume it's the same for all leaves. 

If you really want to be safe, you can use this method:

https://github.com/dreamworksanimation/openvdb/blob/master/openvdb/points/PointDataGrid.h#L232
 
which will fail if the descriptor is *not* identical across all leaf nodes. However, if the source VDBs are all being authored in third-party applications such as Houdini, it's really not necessary because they can't be authored like that anyway.